job-server/src/main/scala/spark/jobserver/JobServer.scala

```diff
@@ -48,49 +51,75 @@ object JobServer {
+    val jobDaoClass = Class.forName(config.getString("spark.jobserver.jobdao"))
+
+    // ensure context-per-jvm is enabled
+    if (sparkMaster.startsWith("yarn") && !contextPerJvm) {
+      throw new InvalidConfiguration("YARN mode requires context-per-jvm")
+    } else if (sparkMaster.startsWith("mesos") && !contextPerJvm) {
+      throw new InvalidConfiguration("Mesos mode requires context-per-jvm")
+    } else if (driverMode == "cluster" && !contextPerJvm) {
+      throw new InvalidConfiguration("Cluster mode requires context-per-jvm")
+    }
+
+    // Check if we are using correct DB backend when context-per-jvm is enabled.
+    // JobFileDAO and H2 mem is not supported.
+    if (contextPerJvm) {
+      if (jobDaoClass.getName == "spark.jobserver.io.JobFileDAO") {
+        throw new InvalidConfiguration("JobFileDAO is not supported with context-per-jvm, use JobSqlDAO.")
+      } else if (jobDaoClass.getName == "spark.jobserver.io.JobSqlDAO" &&
+        config.getString("spark.jobserver.sqldao.jdbc.url").startsWith("jdbc:h2:mem")) {
+        throw new InvalidConfiguration("H2 mem backend is not support with context-per-jvm.")
       }
-      val jobDAO = ctor.newInstance(config).asInstanceOf[JobDAO]
-      val daoActor = system.actorOf(Props(classOf[JobDAOActor], jobDAO), "dao-manager")
-      val dataManager = system.actorOf(DataManagerActor.props(new DataFileDAO(config)), "data-manager")
-      val binManager = system.actorOf(Props(classOf[BinaryManager], daoActor), "binary-manager")
-      val supervisor =
-        system.actorOf(Props(
-          if (contextPerJvm) {
-            classOf[AkkaClusterSupervisorActor]
-          } else {
-            classOf[LocalContextSupervisorActor]
-          },
-          daoActor, dataManager), "context-supervisor")
-      val jobInfo = system.actorOf(Props(classOf[JobInfoActor], jobDAO, supervisor), "job-info")
-
-      // Add initial job JARs, if specified in configuration.
-      storeInitialBinaries(config, binManager)
-
-      // Create initial contexts
-      supervisor ! ContextSupervisor.AddContextsFromConfig
-      new WebApi(system, config, port, binManager, dataManager, supervisor, jobInfo).start()
-    } catch {
-      case e: Exception =>
-        logger.error("Unable to start Spark JobServer: ", e)
-        sys.exit(1)
     }
+
+    // cluster mode requires network base H2 server
+    if (driverMode == "cluster" && jobDaoClass.getName == "spark.jobserver.io.JobSqlDAO") {
+      val jdbcUrl = config.getString("spark.jobserver.sqldao.jdbc.url")
+        if (jdbcUrl.startsWith("jdbc:h2") && !jdbcUrl.startsWith("jdbc:h2:tcp")
+            && !jdbcUrl.startsWith("jdbc:h2:ssl")) {
+          throw new InvalidConfiguration(
+            """H2 backend and cluster mode is not support with file or in-memory storage,
+               use tcp or ssl server.""")
+        }
+    }
+
+    // start embedded H2 server
+    if (config.getBoolean("spark.jobserver.startH2Server")) {
```

Comment:
Dumb question.  Doesn't the startH2Server setting conflict potentially with the sqlDAO setting?  IE it seems for yarn or cluster mode I need to point to a server H2, and it is checked above, but then here we can start an embedded one, so if I use the embedded option what do I set the sqldao setting to and jdbc path?

Reply:
See the cluster docs added. You have to set the jdbc setting to something like `jdbc:h2:tcp://CLUSTER-IP:9092/h2-db;AUTO_RECONNECT=TRUE` to connect to the embedded H2 server.